pkg/relayer/proxy/proxy.go

```diff
@@ -0,0 +1,110 @@
+package proxy
+
+import (
+	"context"
+
+	sdkclient "github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/crypto/keyring"
+	accounttypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+
+	blocktypes "pocket/pkg/client"
+	"pocket/pkg/observable"
+	"pocket/pkg/observable/channel"
+	"pocket/x/service/types"
+	sessiontypes "pocket/x/session/types"
+	suppliertypes "pocket/x/supplier/types"
+)
+
+var (
+	_ RelayerProxy = &relayerProxy{}
+)
+
+type relayerProxy struct {
+	// keyName is the supplier's key name in the keyring. It is used along with the keyring to
+	// get the supplier address and sign the relay responses.
+	keyName string
+	keyring keyring.Keyring
+
+	// blocksClient is the client used to get the latest block height from the blockchain
+	// and be notified about new blocks to update the current session when needed.
+	blocksClient blocktypes.BlockClient
+
+	// accountsQuerier is the querier used to get the application public key from the blockchain,
+	// which is used to verify the relay request signatures.
+	accountsQuerier accounttypes.QueryClient
+
+	// supplierQuerier is the querier used to get the supplier's advertised information from the blockchain,
+	// which contains the supported services, the RPC types, and endpoints needed to query the services.
+	supplierQuerier suppliertypes.QueryClient
+
+	// sessionQuerier is the querier used to get the current session from the blockchain,
+	// which is needed to check if the relay proxy should be serving an incoming relay request.
+	sessionQuerier sessiontypes.QueryClient
+
+	// providedServices is a map of the services provided by the relayer proxy. Each provided service
+	// has the necessary information to start the server that listens for incoming relay requests and
+	// the client that proxies the request to the supported native service.
+	providedServices map[string]ProvidedService
+
+	// servedRelays is an observable that notifies the miner about the relays that have been served.
+	servedRelays observable.Observable[*types.Relay]
```

Comment:
Wdyt of renaming this to `relayServer` or `relayObservable`?

`servedRelays` to me feels like a `read operation` but this is a `subscribe & listen operation`.

Reply:
@bryanchriswhite Curious for your opinion on naming here.

Reply:
It is indeed a read operation. The caller is responsible of subscribing (creating the observer) and listening to its channel.

I should have implemented `ServedRelays` since it's a one-liner which is only returning this property.

Reply:
I still feel like `relayServer` or `relayObservable` is more accurate. Again, not going to push hard on it, but would like another POV @bryanchriswhite

Reply:
**TL;DR, personally, I think `servedRelays` is a good name. :shrug:**

I don't feel like adding the word "observer" adds any real value as it's clear from its type that it is an observable. "Server" doesn't feel right to me either because the observable itself has nothing to do with any network or request handling which is what I would expect from a "server" kind of thing.

My intuition for  naming these has been to put the focus on the thing that the observable is returning as the observable is merely an async primitive, like a promise or future from other languages. I think the name "servedRelays" makes it pretty unambiguously clear what kind of data you're going to get from that observable.

Reply:
Thanks for the response. Let's go with `servedRelays`